File: services/nuldoc/nuldoc-src/pages/PostListPage.ts
import GlobalFooter from "../components/GlobalFooter.ts";
import GlobalHeader from "../components/BlogGlobalHeader.ts";
import PageLayout from "../components/PageLayout.ts";
import Pagination from "../components/Pagination.ts";
import PostPageEntry from "../components/PostPageEntry.ts";
import { Config } from "../config.ts";
import { PostPage } from "../generators/post.ts";
import { elem, Element, h1, header } from "../dom.ts";

export default async function PostListPage(
  posts: PostPage[],
  config: Config,
  currentPage: number,
  totalPages: number,
): Promise<Element> {
  const pageTitle = "投稿一覧";

  const pageInfoSuffix = ` (${currentPage}ページ目)`;
  const metaTitle =
    `${pageTitle}${pageInfoSuffix}｜${config.sites.blog.siteName}`;
  const metaDescription = `投稿した記事の一覧${pageInfoSuffix}`;

  return await PageLayout({
    metaCopyrightYear: config.site.copyrightYear,
    metaDescription,
    metaTitle,
    metaAtomFeedHref: `https://${config.sites.blog.fqdn}/posts/atom.xml`,
    site: "blog",
    config,
    children: elem(
      "body",
      { class: "list" },
      GlobalHeader({ config }),
      elem(
        "main",
        { class: "main" },
        header(
          { class: "page-header" },
          h1({}, pageTitle + pageInfoSuffix),
        ),
        Pagination({ currentPage, totalPages, basePath: "/posts/" }),
        ...posts.map((post) => PostPageEntry({ post, config })),
        Pagination({ currentPage, totalPages, basePath: "/posts/" }),
      ),
      GlobalFooter({ config }),
    ),
  });
}

File: services/nuldoc/lib/nuldoc/pages/post_list_page.rb
module Nuldoc
  module Pages
    class PostListPage
      def initialize(posts:, config:, current_page:, total_pages:)
        @posts = posts
        @config = config
        @current_page = current_page
        @total_pages = total_pages
      end

      def render
        posts = @posts
        config = @config
        current_page = @current_page
        total_pages = @total_pages
        page_title = '投稿一覧'
        page_info_suffix = " (#{current_page}ページ目)"
        meta_title = "#{page_title}#{page_info_suffix}｜#{config.sites.blog.site_name}"
        meta_description = "投稿した記事の一覧#{page_info_suffix}"

        Components::PageLayout.new(
          meta_copyright_year: config.site.copyright_year,
          meta_description: meta_description,
          meta_title: meta_title,
          meta_atom_feed_href: "https://#{config.sites.blog.fqdn}/posts/atom.xml",
          site: 'blog',
          config: config,
          children: DOM::HTMLBuilder.new.build do
            body class: 'list' do
              render Components::BlogGlobalHeader, config: config
              main class: 'main' do
                header(class: 'page-header') { h1 { text "#{page_title}#{page_info_suffix}" } }
                render Components::Pagination, current_page: current_page, total_pages: total_pages,
                                               base_path: '/posts/'
                posts.each { |post| render(Components::PostPageEntry, post: post, config: config) }
                render Components::Pagination, current_page: current_page, total_pages: total_pages,
                                               base_path: '/posts/'
              end
              render Components::GlobalFooter, config: config
            end
          end
        ).build
      end
    end
  end
end
